### PR TITLE
added ability to filter modules per entry point

### PR DESCRIPTION
**Pull request by pjordaan**
with this pull request it is possible to modify the behaviour what files will be bundled in which entry point. It can also be used to filter on a entry point level to avoid redefines. In a sense this way people could write simliar functionality as given by webpack's common chunk plugin.

I also changed the error message in a console.error message and not in a thrown exception.

### CODE REVIEW

File: src/Resources/require.js [hunk 1]
```diff
@@ -132,8 +132,9 @@
     };
 
     window.register = function (name, initializer) {
-        if (_modules[name]) {
-            throw new ModuleRedeclareError(name);
+        if (_modules[name] && typeof console !== 'undefined' && typeof console.error === 'function') {
```
Comment (nicoschoenmaker) [inline src/Resources/require.js]: This could be a separate PR, but
- The a test case at `spec/require.spec.js` should be extended to cover this case
- Right now there if there is a console it keeps the old version of the package, if there is no console it will get the new version.

File: src/Bundler/EntryPoint.php [hunk 2]
```diff
@@ -52,58 +62,29 @@ public function getFile(): File
...
-    {
-        return $this->vendor_files;
-    }
-
-    /**
-     * Return all assets.
-     *
-     * @return File[]
-     */
     public function getAssetFiles(): array
```
Comment (nicoschoenmaker) [inline src/Bundler/EntryPoint.php]: Can this guy be private now?

File: src/Bundler/EntryPoint.php [hunk 3]
```diff
@@ -15,27 +15,37 @@
...
 {
     private $file;
-    private $bundle_files;
-    private $vendor_files;
-    private $asset_files;
+    private $files       = [];
+    private $asset_files = [];
 
-    public function __construct(DependencyNodeInterface $file)
+    public function __construct(DependencyNodeInterface $file, callable $split_strategy = null)
```
Comment (nicoschoenmaker) [inline src/Bundler/EntryPoint.php]: Using generics it would be possible to denote the argument and return type of this callable
```c#
Func<string, string?> splitStrategy
```

PHP can't do this, but we can do something similar if we'd create an interface with one method.

File: src/Bundler/EntryPoint.php [hunk 4]
```diff
@@ -52,58 +62,29 @@ public function getFile(): File
...
      */
-    public function getBundleFiles(): array
+    public function getFilesToBuild(string $output_dir): array
     {
-        return $this->bundle_files;
+        $result = [];
+        foreach ($this->files as $split_file => $dependencies) {
+            $result[$output_dir . '/' . $split_file] = $dependencies;
+        }
+        foreach ($this->getAssetFiles() as $asset_file) {
```
Comment (yannickl88) [inline src/Bundler/EntryPoint.php]: you can use the property instead of the getter.

Comment (pjordaan) [reply src/Bundler/EntryPoint.php]: in the updated pr the getter no longer exists

File: src/Bundler/EntryPoint.php [hunk 5]
```diff
@@ -15,27 +17,37 @@
...
+            $split_strategy = new OneOnOneSplittingStrategy();
+        }
         $this->file = $file->getFile();
 
-        // Split the input files into bundle and vendor files.
-        $this->bundle_files = [$file];
-        $this->vendor_files = [];
-        $this->asset_files  = [];
+        $result = $split_strategy->resolveChunk($file->getFile()->getName(), $file);
+        if ($result) {
```
Comment (yannickl88) [inline src/Bundler/EntryPoint.php]: `null !== $result` for clarity?

File: src/Bundler/EntryPoint.php [hunk 6]
```diff
@@ -52,58 +64,19 @@ public function getFile(): File
...
-    public function getVendorFile(string $output_dir): File
+    public function getFilesToBuild(string $output_dir): array
     {
-        return new File($output_dir . '/' . sprintf('%s.vendor.js', $this->file->getBaseName()));
+        $result = [];
+        foreach ($this->files as $split_file => $dependencies) {
+            $result[$output_dir . '/' . $split_file] = $dependencies;
+        }
+        foreach ($this->asset_files as $asset_file) {
+            $result[$output_dir . '/' . $asset_file->getFile()->getName()] = [$asset_file];
```
Comment (yannickl88) [inline src/Bundler/EntryPoint.php]: `getFile()` can also return a `Module`. These do not have an extension in the name. Maybe use `$asset_file->getFile()->path`.

File: src/Config/ConfigInterface.php [hunk 7]
```diff
@@ -25,6 +26,13 @@
      */
     public function isDev(): bool;
 
+    /**
+     * Returns a class that resolves chunk points into which output file it should be sent to.
+     *
+     * @return callable
```
Comment (yannickl88) [inline src/Config/ConfigInterface.php]: `@return EntryPointSplittingStrategyInterface`

File: src/Resources/require.js [hunk 8]
```diff
@@ -133,7 +133,10 @@
 
     window.register = function (name, initializer) {
         if (_modules[name]) {
-            throw new ModuleRedeclareError(name);
+            if (typeof console !== 'undefined' && typeof console.error === 'function') {
+                console.error(new ModuleRedeclareError(name));
```
Comment (yannickl88) [inline src/Resources/require.js]: are we sure we want to log it as an error? Maybe a warning might be more appropriate.

File: src/Split/EntryPointSplittingStrategyInterface.php [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<?php
+/**
+ * @copyright 2018 Hostnet B.V.
+ */
+declare(strict_types=1);
+namespace Hostnet\Component\Resolver\Split;
+
+use Hostnet\Component\Resolver\Import\DependencyNodeInterface;
+
+interface EntryPointSplittingStrategyInterface
```
Comment (yannickl88) [inline src/Split/EntryPointSplittingStrategyInterface.php]: Can you add some doc what the purpose of implementations is.

File: src/Bundler/EntryPoint.php (hunk 5 above)
Comment (pjordaan) [reply src/Bundler/EntryPoint.php]: I actually removed that from a previous commit for the reason that you can't have an empty file name, so it seems to me that you shouldn't be able to split a chunk to an empty file. I could have added an error on empty string or just take the simple way out and consider null and '' as 'do not include'

File: src/Resources/require.js (hunk 8 above)
Comment (pjordaan) [reply src/Resources/require.js]: console.warn is fine to me. I would still keep it as an Error instance sent to the console as all browser will display a stack trace with the error message giving you a better dev experience on debugging.

File: src/Config/ConfigInterface.php (hunk 7 above)
Comment (nicoschoenmaker) [inline src/Config/ConfigInterface.php]: `@return EntryPointSplittingStrategyInterface`

File: src/Config/SimpleConfig.php [hunk 10]
```diff
@@ -74,6 +79,11 @@ public function __construct(
         $this->reporter           = $reporter ?? new NullReporter();
     }
 
+    public function getSplitStrategy(): EntryPointSplittingStrategyInterface
```
Comment (nicoschoenmaker) [inline src/Config/SimpleConfig.php]: Missing test for `getSplitStrategy`
